### ISSUE TITLE
cert-dcl58-cpp flags legitimate specializations (false positive)

### ISSUE THREAD

**Issue by LegalizeAdulthood** [llvm/llvm-project#45454]
|  |  |
| --- | --- |
| Bugzilla Link | [46109](https://llvm.org/bz46109) |
| Version | unspecified |
| OS | Windows NT |
| CC | @EugeneZelenko,@johnmcfarlane,@Xazax-hun |

## Extended Description 
You're allowed to inject specializations of std::hash into namespace std for your own types.

https://en.cppreference.com/w/cpp/utility/hash

**Comment by LegalizeAdulthood** [comment]
The issue no longer reproduces in clang-tidy 14, so I'm closing.

**Comment by Febbe** [comment]
Please reopen, it still fails (15.0-rc2) for templated objects:

```
template <typename T>
struct MyStruct{};

namespace std {
template <typename T>
struct hash<MyStruct<T>> {
    auto operator()(const MyStruct& k) const noexcept -> std::size_t {
        return 1U;
    }
};
}    // namespace std

```

**Comment by captaincrutches** [comment]
I can confirm @EugeneZelenko's case, I'm also getting that warning on a partial specialization as of 15.0.0.

It seems to have started in 15 actually, I wasn't getting it in 14.

There were [changes made](https://reviews.llvm.org/D23421?id=88541) to this warning in 15, seemingly with the goal of allowing full specializations, which _do_ work - but partial ones are throwing out the false positive now.

**Comment by jbytheway** [comment]
Just giving this a bump because I've run into the same issue of false positives on partial specializations on version 16.0.4.

**Comment by N-Dekker** [comment]
Just ran into the same issue (I think), trying to add structured binding support to a class template, `itk::ImageRegion<unsigned int>`, by specializing `std::tuple_size` and `std::tuple_element` for `itk::ImageRegion`: https://github.com/InsightSoftwareConsortium/ITK/pull/4367

Affected code:

```cpp
namespace std
{
template <unsigned int VImageDimension>
struct tuple_size<itk::ImageRegion<VImageDimension>> : std::integral_constant<size_t, 2>
{};

template <size_t VTupleIndex, unsigned int VImageDimension>
struct tuple_element<VTupleIndex, itk::ImageRegion<VImageDimension>>
  : conditional<VTupleIndex == 0, itk::Index<VImageDimension>, itk::Size<VImageDimension>>
{};

}
```

Clang Tidy output from Visual Studio 2019, Clang Power Tools 2023.9.0, LLVM 17.0.1:

    warning: modification of 'std' namespace can result in undefined behavior [cert-dcl58-cpp]
         | struct tuple_size<itk::ImageRegion<VImageDimension>> : std::integral_constant<size_t, 2>
         |        ^
...

    warning: modification of 'std' namespace can result in undefined behavior [cert-dcl58-cpp]
         | struct tuple_element<VTupleIndex, itk::ImageRegion<VImageDimension>>
         |        ^

Hope it can still be fixed. Otherwise, do you have a suggestion how to work around this warning, in user code?

**Comment by Febbe** [comment]
> Hope it can still be fixed. Otherwise, do you have a suggestion how to work around this warning, in user code?

It's a bug in the checkers code. So if someone has time, he could always write a fix and open a PR.
There is no work around / other way to do things like that. The code is completely correct. I would just disable the warning either per line, or in the .clang-tidy config.

**Comment by PiotrZSL** [comment]
Currently //NOLINT(cert-dcl58-cpp) or //NOLINTNEXTLINE(cert-dcl58-cpp) as an workaround

**Comment by csisy** [comment]
> Please reopen, it still fails (15.0-rc2) for templated objects:
> 
> ```
> template <typename T>
> struct MyStruct{};
> 
> namespace std {
> template <typename T>
> struct hash<MyStruct<T>> {
>     auto operator()(const MyStruct<T>& k) const noexcept -> std::size_t {
>         return 1U;
>     }
> };
> }    // namespace std
> ```

I have the same scenario (templated class cannot be used in specialization). For now, I needed to turn this off.
